pkg/provisioner/provisioner.go

```diff
@@ -0,0 +1,273 @@
+package provisioner
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+
+	"github.com/iver-wharf/wharf-core/pkg/logger"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+)
+
+var log = logger.NewScoped("PROVISIONER")
+
+var podInitCloneArgs = []string{"git", "clone"}
+var podContainerListArgs = []string{"/bin/sh", "-c", "ls -alh"}
+
+// Provisioner is an interface declaring what methods are required
+// for a provisioner.
+type Provisioner interface {
+	Serve(ctx context.Context) error
+	ListPods(ctx context.Context) error
+	DeletePod(ctx context.Context, UID string) error
+}
+
+type k8sProvisioner struct {
+	Namespace  string
+	Clientset  *kubernetes.Clientset
+	Pods       corev1.PodInterface
+	restConfig *rest.Config
+	events     corev1.EventInterface
+}
+
+// NewK8sProvisioner returns a new provisioner implementation that targets
+// Kubernetes using a specific Kubernetes namespace and REST config.
+func NewK8sProvisioner(namespace string, restConfig *rest.Config) (Provisioner, error) {
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	return k8sProvisioner{
+		Namespace:  namespace,
+		Clientset:  clientset,
+		Pods:       clientset.CoreV1().Pods(namespace),
+		restConfig: restConfig,
+		events:     clientset.CoreV1().Events(namespace),
+	}, nil
+}
+
+func (p k8sProvisioner) ListPods(ctx context.Context) error {
+	podList, err := p.Pods.List(ctx, metav1.ListOptions{
+		LabelSelector: "app.kubernetes.io/name=wharf-cmd-worker,app.kubernetes.io/managed-by=wharf-cmd-provisioner,wharf.iver.com/instance=prod",
+	})
+	if err != nil {
+		return err
+	}
+	pods := podList.Items
+
+	printPod := func(pod metav1.ObjectMeta, i int) func(logger.Event) logger.Event {
+		return func(ev logger.Event) logger.Event {
+			ev = ev.WithInt("index", i).
+				WithString("UID", string(pod.UID)).
+				WithString("name", pod.Name).
+				WithString("namespace", pod.Namespace)
+
+			for k, v := range pod.Labels {
+				ev = ev.WithString(k, v)
+			}
+
+			return ev
+		}
+	}
+
+	log.Info().WithInt("count", len(pods)).Message("Fetched pods with matching labels.")
+	for i, pod := range pods {
+		log.Info().WithFunc(printPod(pod.ObjectMeta, i)).Message("Pod")
+	}
+
+	return nil
+}
+
+func (p k8sProvisioner) DeletePod(ctx context.Context, UID string) error {
+	podList, err := p.Pods.List(ctx, metav1.ListOptions{
+		LabelSelector: "app.kubernetes.io/name=wharf-cmd-worker,app.kubernetes.io/managed-by=wharf-cmd-provisioner,wharf.iver.com/instance=prod",
```

Comment:
If this is the exact same as in ListPods then could you extract it into a global const or var?